ENH:  routine to estimate the noise standard deviation from an image

grlee77 opened the pull request:
This PR adds a routine for estimating the (Gaussian) noise standard deviation from an image.  This should be of use with the various denoising routines in the usual case when the underlying noise standard deviation is unknown.

The algorithm implemented is based on the median absolute deviation of wavelet detail coefficients, so it requires `PyWavelets` to run.  

For example, this could be used by the method of #1833 to automate the threshold estimation.

A minimal example:
```python
import skimage.data
from skimage import img_as_float
from skimage.restoration import estimate_sigma
img = img_as_float(skimage.data.camera())
sigma = 0.1
img = img + sigma * np.random.standard_normal(img.shape)
sigma_hat = estimate_sigma(img, multichannel=False)
print("True sigma = {}".format(sigma))
print("Estimated sigma = {}".format(sigma_hat))
```
```
True sigma = 0.1
Estimated sigma = 0.09989763880477982
```

grlee77 commented:
I left `multichannel` a required argument, because I thought defaulting to False was a bit dangerous.  If a color image is input, but `multichannel` False, this will cause a substantial error in the estimation of sigma.  (This is because the DWT is then improperly performed along the channels dimension as well).

stefanv commented:
Perhaps @arokem would be willing to help us review this one.  He has dived deep into the landscape of noise estimation for neuroimaging in the past.

arokem commented:
Yeah, I can take a look.

grlee77 commented:
The error on Travis is due to the presence of the docstring example, but lack of PyWavelets so the doctest fails.  Is there a simple way to tell the testing to skip it?

grlee77 commented:
The neuroimaging use case (at least for MRI) is a bit more complicated and would benefit from a more specialized solution.

Often, one only has magnitude images which have Rician rather than Gaussian noise.  Gaussian is a reasonable approximation when SNR is not too low, though.  Also, for MRI, the noise standard deviation is often not spatially uniform so it is better to have localized estimates of sigma rather than a single value for the whole image/volume.

stefanv commented:
Try ``# doctest: +SKIP``

grlee77 commented:
Okay, the doctest is skipped now.  It fails now because I listed PyWavelets in the `optional_dependencies.rst` file, but it is not available.

stefanv commented on skimage/restoration/_denoise.py:
```diff
@@ -355,3 +361,93 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
+
+def estimate_sigma(im, multichannel, average_sigmas=True):
+    """
+    Robust wavelet-based estimator of the noise standard deviation [1]_.
```
No references in first line of docstring.

stefanv commented:
Failing because of docstring formatting.

JDWarner commented:
Once we officially tag and release 0.12, we should seriously discuss adding PyWavelets as a required dependency for 0.13+. Then the docstring will pass, instead of needing to be skipped.

stefanv commented:
Why don't you file an issue with 0.13 as the milestone, Josh?
​

grlee77 commented:
The remaining failure (segfault!) on Python 2.6 does not seem related to this PR:

```
skimage.feature.tests.test_util.test_plot_matches ... tools/travis_script.sh: line 111: 10304 Segmentation fault      (core dumped) nosetests $TEST_ARGS

The command "tools/travis_script.sh" exited with 139.
```

arokem commented on skimage/restoration/_denoise.py:
```diff
@@ -355,3 +361,98 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
+    .. [1] Donoho, David L., and Jain M. Johnstone. "Ideal spatial adaptation
+       by wavelet shrinkage." Biometrika 81.3 (1994): 425-455.
+    """
+    detail_coeffs = np.asarray(detail_coeffs)
```
Is the call to `asarray` a standard thing in skimage, even when input is defined to be an array in the docstring?

arokem commented on skimage/restoration/_denoise.py:
```diff
@@ -355,3 +361,98 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
+    detail_coeffs = np.asarray(detail_coeffs)
+    if distribution.lower() == 'gaussian':
+        # 75th quantile of the underlying, symmetric noise distribution:
+        # denom = scipy.stats.norm.ppf(0.75)
```
Out of curiosity, why are you not actually doing the call to scipy.stats?

grlee77 replied:
I put that there as a comment to remind myself where the value of `0.6745` in the Donoho paper came from (I did not see it explained within the reference), but figured there was no need to recompute it each time.

arokem replied:
Sure - that makes sense.

arokem commented on skimage/restoration/_denoise.py:
```diff
@@ -355,3 +361,93 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
+    Returns
+    -------
+    sigma : float
+        estimated noise standard deviation.  See [1]_
```
You might want to tell people to look at section 4.2 of the paper. It's rather long. Also: period at the end of this sentence.

arokem commented:
Overall - looks good to me. Once the small comments are addressed, I think this is good to go.

matthew-brett commented on skimage/restoration/_denoise.py:
```diff
@@ -355,3 +361,98 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
+
+    References
+    ----------
+    .. [1] Donoho, David L., and Jain M. Johnstone. "Ideal spatial adaptation
```
Typo "Jain" should be "Iain".

grlee77 commented:
Should `PyWavelets` be listed in `optional_dependencies.rst`?  It seemed like doing so caused Travis to try and find wheels for PyWavelets at http://travis-wheels.scikit-image.org/.  There is support for `bdist_wheel` in `PyWavelets`, so I think it should be possible to generate the necessary wheels fairly easily.

I have tested with `PyWavelets` v0.3.0 and from current master.  For a 512x512 grayscale image, `estimate_sigma` requires ~100 ms on v0.3.0 and ~10 ms for master.

grlee77 commented:
Also,  should I add a conversion via `img_as_float` for the input to `estimate_sigma`?  The estimate_sigma function will work fine for integer images (`PyWavelets` will convert to double precision internally in that case), but it appears the other denoising functions convert their inputs to float, which could potentially lead to users calling `estimate_sigma` on the integer image and then passing that sigma on to another function where it would be much too large once the image gets rescaled to [0, 1].

JDWarner commented:
@grlee77 Realistically, we aren't merging this before 0.12 is released. My thought process moving forward is to make PyWavelets a required dependency and open the floodgates to wavelet-based algorithms as major new features in 0.13. Further discussion on this in #1838. We're not talking a long time, here - work on 0.13dev should begin soon.

For now, it would probably cleanest (but not entirely satisfying) to continue with this PR assuming PyWavelets will be made a dependency in a separate PR. No optional checks or warnings, just assume it'll be there. Yes, the tests will fail until after 0.12 is tagged and released, but then a PR making PyWavelets a dependency can be merged and a simple rebase will both bring it into the green. You can move forward with development/testing locally in virtualenvs.

emmanuelle commented:
@blink1073 any idea about the segmentation fault of Travis?

blink1073 commented:
@emmanuelle, no idea, but it looks like it was unrelated to this PR and caused by matplotlib.  I restarted the build to see if it is reproducible.

blink1073 commented:
Ah, as of 21 December, we're using matplotlib 1.5 in our builds.  http://travis-wheels.scikit-image.org/.

Matplotlib 1.5 does not support Python 2.6.  We may have to follow suit for v0.12.  
https://github.com/matplotlib/matplotlib/blob/master/setup.py#L121

@scikit-image/core, thoughts?

emmanuelle commented:
@blink1073 do we need matplotlib 1.5 for the build? python 2.6 should be deprecated only after 0.12.

blink1073 commented:
We could pin to a lower version of matplotlib on our Travis build, and just rely on the fact that the user shouldn't be installing a newer version in they are on Py26.

emmanuelle commented:
sounds good

blink1073 commented:
See https://github.com/scikit-image/scikit-image/pull/1841

grlee77 commented:
@JDWarner:  That all sounds good to me.  I did not expect it to go in for 0.12.  I can eventually rebase this to remove some of the unecessary commits adding and removing the optional dependency, etc.

sciunto commented:
@grlee77 I think it would be good to give a last round to this PR. Pywavelet and denoising are integrated to skimage now. Can I ask you to rebase this PR, please?

sciunto commented on skimage/restoration/_denoise.py:
```diff
@@ -4,6 +4,12 @@
 from ..restoration._denoise_cy import _denoise_bilateral, _denoise_tv_bregman
 from .._shared.utils import _mode_deprecations
 
+try:
```
try except can be removed.

sciunto commented on skimage/restoration/_denoise.py:
```diff
@@ -355,3 +361,97 @@ def denoise_tv_chambolle(im, weight=50, eps=2.e-4, n_iter_max=200,
+    >>> img = img_as_float(skimage.data.camera())
+    >>> sigma = 0.1
+    >>> img = img + sigma * np.random.standard_normal(img.shape)
+    >>> sigma_hat = estimate_sigma(img, multichannel=False)  # doctest: +SKIP
```
Do we need to skip?

sciunto commented on skimage/restoration/tests/test_denoise.py:
```diff
@@ -220,5 +220,55 @@ def test_no_denoising_for_small_h():
     assert np.allclose(denoised, img)
 
 
+@dec.skipif(not restoration._denoise.pywt_available)
```
I think these decorators can be removed too.

grlee77 commented:
Thanks @sciunto.  I have rebased on master and removed the checks for `pywt`.

I should probably also add an example or modify one of the existing ones to make use of this estimation.  I can either do that here or in a separate PR.

grlee77 commented:
The Travis failure is during the Documentation build and appears to be unrelated (a duplicate reference related to seam carving).
```
[R411]_']:52: WARNING: duplicate citation R411, other instance in /home/travis/build/scikit-image/scikit-image/doc/source/api/skimage.util.rst
```

jni commented:
@grlee77 the documentation issue is known but we don't have any insight into it...

https://github.com/scikit-image/scikit-image/issues/2237

sciunto commented on skimage/restoration/_denoise.py:
```diff
@@ -470,3 +470,94 @@ def denoise_wavelet(img, sigma=None, wavelet='db1', mode='soft'):
+    if distribution.lower() == 'gaussian':
+        # 75th quantile of the underlying, symmetric noise distribution:
+        # denom = scipy.stats.norm.ppf(0.75)
+        denom = 0.67448975019608171
```
It looks you hesitate between scipy or a hardcoded value. I'm in favour of scipy with `from scipy import stats`

sciunto commented on skimage/restoration/_denoise.py:
```diff
@@ -470,3 +470,94 @@ def denoise_wavelet(img, sigma=None, wavelet='db1', mode='soft'):
+    References
+    ----------
+    .. [1] D. L. Donoho and I. M. Johnstone. "Ideal spatial adaptation
+       by wavelet shrinkage." Biometrika 81.3 (1994): 425-455.
```
Could you add `DOI:10.1093/biomet/81.3.425` please?

sciunto commented on skimage/restoration/_denoise.py:
```diff
@@ -470,3 +470,94 @@ def denoise_wavelet(img, sigma=None, wavelet='db1', mode='soft'):
+    >>> img = img + sigma * np.random.standard_normal(img.shape)
+    >>> sigma_hat = estimate_sigma(img, multichannel=False)
+    """
+    if multichannel:
```
What would happen if I load a data.camera() and specify multichannel=True?
I think you need to check if the image first and raise an exception if the parameters and the data are incompatible

sciunto replied:
and you can add a test to cover this :)

jni replied:
I would consider *not* requiring this. Sanity checks are generally useful, but you can spend 90% of your time doing them if you're not careful. In this case, the user has to *explicitly* set a kwarg to generate the error, so it falls under "fair to fail" imho.

I certainly wouldn't fight such a check, but I'm not sure it's required either. I'll leave it up to you two to argue either way, just wanted to provide an alternative!

grlee77 replied:
@sciunto.  How would you propose checking?  I don't think we should assume multichannel means RGB or RGBA.  For example, data from hyperspectral imaging could have different number of channels.  I would prefer to err on the side of user control as long as there is a valid use case (even if that means the user could potentially shoot themselves in the foot!).

sciunto commented on skimage/restoration/_denoise.py:
```diff
@@ -470,3 +470,94 @@ def denoise_wavelet(img, sigma=None, wavelet='db1', mode='soft'):
+    ----------
+    im : ndarray
+        Image for which to estimate the noise standard deviation.
+    multichannel : bool
```
I think I would make it optional and set to False by default.

grlee77 replied:
I agree that it is a bit inconvenient to have to type out multichannel=False every time when working with graycale images.
That said, I expect that a fairly common error would then be passing in a color image, but not setting the multichannel argument.  Maybe raise a warning if the size is 3 on the last axis and multichannel is False?

sciunto commented on skimage/restoration/_denoise.py:
```diff
@@ -470,3 +470,94 @@ def denoise_wavelet(img, sigma=None, wavelet='db1', mode='soft'):
+        Image for which to estimate the noise standard deviation.
+    multichannel : bool
+        Estimate sigma separately for each channel.
+    average_sigmas : bool, optional
```
And this one, False by default.

sciunto commented on skimage/restoration/_denoise.py:
```diff
@@ -470,3 +470,94 @@ def denoise_wavelet(img, sigma=None, wavelet='db1', mode='soft'):
+    Notes
+    -----
+    The estimation algorithm is based on the median absolute deviation of the
+    wavelet detail coefficients as described in section 4.2 of [1]_.
```
Maybe you can say a word on the assumption of a gaussian noise.

sciunto commented on skimage/restoration/_denoise.py:
```diff
@@ -470,3 +470,94 @@ def denoise_wavelet(img, sigma=None, wavelet='db1', mode='soft'):
+    References
+    ----------
+    .. [1] D. L. Donoho and I. M. Johnstone. "Ideal spatial adaptation
+       by wavelet shrinkage." Biometrika 81.3 (1994): 425-455.
```
Add also the DOI here please.

sciunto commented:
@grlee77 Thanks for pushing this :) very appreciated. I looked more carefully to your code. I think I found a possible bug and I raise few minor stuffs. Otherwise, it looks good. I'll look at it again once you fixed these little things.

Thanks.

sciunto commented:
Also, if you can incorporate this in the example, that would be great! ;)

jni commented:
LGTM, thanks @grlee77!

soupault commented:
Awesome! Thanks a lot @grlee77!
